### PR TITLE
Add few shot and multiple choice to ICL evaluation

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?
This PR extends the existing ICL-LM evaluator framework to support few shot and multiple choice (e.g. PIQA)

I create a custom multiple choice in-context learning data loader, a special NLP metric of ICL-multiple choice, and extend the existing ICL-LM data loader to support few shot prompting.

Manually evaluation GPT Neo confirms Lambada 57.23% and PIQA 71%

# What issue(s) does this change relate to?

# Before submitting
- [x ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x ] Did you update any related docs and document your change?
- [x ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x ] Did you run the tests locally to make sure they pass?
- [ x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:217]: could you comment explaining the `choice_groupings` list? That one in particular took me a bit to understand that it was indexing into `continuation_indices` because each choice is its own continuation.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:285]: Please either use a logger or remove the print statement

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:337]: Please either use a logger or remove the print statement

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:282]: since this is an eval dataset, should shuffle be false?

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:336]: same as above on shuffle

File: tests/datasets/test_in_context_learning_datasets.py
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:33]: I think it may be worth it to actually write out the expected value of each of these keys for both of these dataloader tests, for the first example. otherwise they are just testing that code doesn't crash and that the dictionary returned is correctly formed. As the dataset got more complicated with fewshot, I think this became more necessary.

File: tests/datasets/test_in_context_learning_datasets.py
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:94]: Is there a need to parametrize over the two datasets? They look of the same format to me.

File: composer/metrics/nlp.py
Comment (dakinggg) [inline composer/metrics/nlp.py:291]: I think this was added to the docstring by mistake?

**Review by dakinggg — APPROVED** [review]
Approving to unblock since I will be gone for the next week, but left a bunch of comments.

**Comment by mvpatel2000** [comment]
Same as https://github.com/mosaicml/composer/pull/1879, please ensure that new tests are not slower than 5s on CPU/GPU!

File: composer/metrics/nlp.py
Comment (bmosaicml) [reply composer/metrics/nlp.py:286]: Good catch, this was a merging mistake

File: tests/datasets/test_in_context_learning_datasets.py
Comment (bmosaicml) [reply tests/datasets/test_in_context_learning_datasets.py]: They test different number of answers per Q (PIQA has two choices, hellaswag 4)

File: composer/datasets/in_context_learning_evaluation.py
Comment (abhi-mosaic) [inline composer/datasets/in_context_learning_evaluation.py:352]: I don't think this is quite right, like I think `attention_mask` is meant to be `1` where there are tokens and `0` where there is padding. I think the `eos_token_id` is an artifact of the GPT2Tokenizer but not a general thing.

I think the safest thing todo here is to copy what HF's default data collator does (maybe even import and use directly!): https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/data/data_collator.py#L1152-L1193

Comment (abhi-mosaic) [inline composer/datasets/in_context_learning_evaluation.py:352]: I don't think this is quite right, like I think `attention_mask` is meant to be `1` where there are tokens and `0` where there is padding. I think the `eos_token_id` is an artifact of the GPT2Tokenizer but not a general thing.

I think the safest thing todo here is to copy what HF's default data collator does (maybe even import and use directly!): https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/data/data_collator.py#L1152-L1193

File: composer/trainer/trainer.py
Comment (abhi-mosaic) [inline composer/trainer/trainer.py:2554]: not sure if this is the linter acting up but probably best to `git checkout dev -- trainer/trainer.py` if no changes intended.

File: composer/datasets/in_context_learning_evaluation.py
Comment (abhi-mosaic) [inline composer/datasets/in_context_learning_evaluation.py:328]: Related to below, is there a reason we are padding with the `eos_tok_id` rather than the `pad_tok_id`? I guess if it's consistent all the way through that's fine, but it might be good to add a comment about it.

File: composer/datasets/in_context_learning_evaluation.py
Comment (abhi-mosaic) [inline composer/datasets/in_context_learning_evaluation.py:359]: Are these functions necessary? I thought Composer could handle `get_num_samples_in_batch` HF style batches automatically.

The `update_metric` doesn't seem to be used.

File: composer/datasets/in_context_learning_evaluation.py
Comment (abhi-mosaic) [inline composer/datasets/in_context_learning_evaluation.py:439]: Related to above, if we can get away with Composer's default `get_num_samples_in_batch` function, then we can skip the `DataSpec` wrapper entirely and just return a `DataLoader`. But ofc if the custom function is needed, then no changes here.

**Review by abhi-mosaic — COMMENTED** [review]
Looking good! I added some comments about `eos` padding and some (maybe) superfluous functions. @bmosaicml I think it would also be good to get 1 more NLP team reveiwer who is more familiar with eval to look over the tests... I am not an expert here so I could have missed something.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:359]: I'm guessing that the default ends up raising here https://github.com/mosaicml/composer/blob/c99230435cd23728e9e5d8cbdc62bc92cbbaa64d/composer/core/data_spec.py#L297 or maybe even crashing here (https://github.com/mosaicml/composer/blob/c99230435cd23728e9e5d8cbdc62bc92cbbaa64d/composer/core/data_spec.py#L292) because of the non tensor items in the dictionary. Possible the default should get slightly smarter (i.e. look for `input_ids`, or skip non tensor items in the batch). Made a JIRA to look into this https://mosaicml.atlassian.net/browse/CO-1660.

Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:359]: `update_metric` is what composer calls to update metrics. It isn't implemented on the base class (maybe it should be?), so any subclass of `ComposerModel` that wants to do eval needs to implement it.

File: composer/datasets/in_context_learning_evaluation.py
Comment (bmosaicml) [reply composer/datasets/in_context_learning_evaluation.py:352]: I think their collator is more or less the same logic, where they set the attention mask to be equal to the places where they did/didn't insert padding tokens.

I inserted the eos_tok_id tokens as padding in my collation function. As long as the eos tok is never used to have real semantic meaning then I think we would be good.

File: composer/datasets/in_context_learning_evaluation.py
Comment (bmosaicml) [reply composer/datasets/in_context_learning_evaluation.py:359]: yep you're right about the update metric function being superfluous and needing to delete it. Also Daniel is right about why the default doesn't work!

File: composer/datasets/in_context_learning_evaluation.py
Comment (bmosaicml) [reply composer/datasets/in_context_learning_evaluation.py:328]: The actual token used by the data loader can be the pad token or the eos token, so for clarity I will change the internal naming to pad token. For the purposes of constructing the data loader, the GPTTokenizer's eos_tok_id is the last id in the GPT embedding, the pad token is the index after, and does not appear in the embedding, hence we can't actually pass in the pad_token_id in our tensors

**Review by dakinggg — APPROVED** [review]
LGTM, maybe would be good to have another NLP person have a glance. Also trusting that you and abhi worked out that attention mask thing, I didn't look at that.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:426]: as discussed offline, for MC, suggest overriding the split batch function with something that raises an error, to make absolutely sure that batches don't get split in a way your code doesn't expect. Another option is to restructure the batch so that the second dimension is actually splittable, but we can defer that since its more complicated.

Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:426]: https://mosaicml.atlassian.net/browse/CO-1694

File: composer/models/huggingface.py
Comment (bmosaicml) [reply composer/models/huggingface.py:277]: yeah basically, if the labels are part of the batch when you call forward, then the HF model will calculate cross entropy loss, which is a waste of time for the ICL eval metrics

Comment (dakinggg) [reply composer/models/huggingface.py:277]: Oh, I see. I think this might break some other functionality though, because other uses might assume that the output will contain the loss if `self.use_logits` is `False`. do you think it would make sense to only pop the labels before calling `forward` `if self.use_logits or batch.get('mode', None) == 'icl_task':`?

File: composer/models/huggingface.py
Comment (dakinggg) [inline composer/models/huggingface.py:45]: looks like a bunch of stuff got deleted in a bad merge

File: composer/metrics/nlp.py
Comment (alextrott16) [inline composer/metrics/nlp.py:335]: Same as above

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py]: double import

File: composer/models/huggingface.py
Comment (dakinggg) [inline composer/models/huggingface.py:305]: I think we don't want to pop the labels in the else case here? to match the previous behavior

File: composer/models/huggingface.py
Comment (dakinggg) [inline composer/models/huggingface.py:306]: lm_task -> icl_task?

File: composer/models/huggingface.py
Comment (dakinggg) [inline composer/models/huggingface.py:292]: I think this might actually cause a problem. If you pass the batch through without "labels", the logits are no longer at index one in the tuple. Still verifying, just want to block merging until we figure it out.

File: composer/models/huggingface.py
Comment (dakinggg) [inline composer/models/huggingface.py:306]: can you just move this code back to where it was? you can still change the conditional to be this or statement with icl task

File: composer/models/huggingface.py
Comment (dakinggg) [reply composer/models/huggingface.py:292]: ok, so we need to change
```
            if self.config.use_return_dict:
                output = output['logits']
            else:
                # logits are at index 1 in the output tuple
                output = output[1]
```
to
```
            if self.config.use_return_dict:
                output = output['logits']
            else:
                # logits are at index 0 because loss is not present
                output = output[0]
```

File: composer/models/huggingface.py
Comment (dakinggg) [inline composer/models/huggingface.py:297]: Also need to update the comment to be true

File: composer/models/huggingface.py
Comment (dakinggg) [reply composer/models/huggingface.py:305]: I think you added this back in somehow?

**Review by dakinggg — APPROVED** [review]
LGTM once you resolve the last open conversations. Could you please also put in the PR description that numbers that you get from a manual test with a pretrained model + the full dataset?

**Review by bmosaicml — COMMENTED** [review]
lgtm!

File: setup.py
Comment (mvpatel2000) [inline setup.py:172]:
```suggestion
extra_deps['nlp'] = [
    'transformers>=4.11,<5',
    'datasets>=2.4,<3',
]
```

Comment (mvpatel2000) [reply setup.py:172]: nit: to be consistent with style elsewhere

File: composer/metrics/__init__.py
Comment (mvpatel2000) [inline composer/metrics/__init__.py:15]: nit: can we add a comma and go back to old format to be consistent please

**Review by mvpatel2000 — COMMENTED** [review]
One more small nit: can we change from JSONL to JSON to clearly indicate file size limit (which applies bc we use Github)

A few small nits... i'll stamp once we figure out release process

File: setup.py
Comment (dakinggg) [reply setup.py:172]: this was yapf

Comment (mvpatel2000) [reply setup.py:172]: Probably because there was no trailing comma

File: composer/metrics/__init__.py
Comment (bmosaicml) [reply composer/metrics/__init__.py:15]: what do you mean by "add a comma"?

Comment (mvpatel2000) [reply composer/metrics/__init__.py:15]: add a trailing comma behind `'InContextLearningMetric'`

File: composer/metrics/nlp.py
Comment (knighton) [inline composer/metrics/nlp.py:307]: should this be one blank line?

**Review by knighton — APPROVED** [review]

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:205]: this trailing whitespace is causing lint to fail

**Pull request merged**
